ipamd assigns IPs from detached ENIs

I've been debugging an issue where ipamd assigns IPs from ENIs that are not attached to the EC2 instance.  This is support case# 6601452261 (I dunno if you have access to that).

I've tracked it down to the ENI reconciliation logic.  ipamd starts out with 3 ENIs, and assigns IPs from these ENIs, and all is good.  Then one of the ENIs gets detached, outside of ipamd's control.  Here's where badness starts; the datastore will continually refuse to remove the ENI, since some of its IPs are assigned.  And ipamd will continue to assign IPs from this detached ENI, meaning that our pods end up with IPs that simply don't work.

In our specific case, we still don't know why the ENI gets detached; it seems to happen shortly after our VM reboots.  We also don't know why our VM is rebooting.  The working theory is that our EC2 instances end up with too many combined ENI+EBS attachments; i.e. more than the limit for our instance type.  Perhaps that results in VM reboot, and subsequent ENI detachment.

Regardless of the reason, it seems clear that ipamd shouldn't continue to assign IPs from detached ENIs.  I have a PR for a fix that I'll send shortly.  Below are logs that demonstrate the problem from v1.5.5, that I've annotated whilst debugging.

```
***** 0) ipamd startup after reboot
2019-11-20T16:47:07.325Z [INFO] Starting L-IPAMD v1.5.5  ...

***** 1) ipamd discovers 3 ENIs, including eni-02797faf7a67b1590
2019-11-20T16:47:37.354Z [DEBUG]        Total number of interfaces found: 3
2019-11-20T16:47:37.356Z [DEBUG]        Found ENI: eni-02797faf7a67b1590, MAC 06:48:a2:a8:91:f0, device 3
2019-11-20T16:47:37.358Z [DEBUG]        Found ENI: eni-0ea92c15e054d06d3, MAC 06:91:7d:d1:10:9c, device 2
2019-11-20T16:47:37.360Z [DEBUG]        Found ENI: eni-0578f3df9b86e409c, MAC 06:f1:59:05:99:ac, device 0

***** 2) ipamd reconciles 3 ENIs from AWS local metadata, including eni-02797faf7a67b1590 with IP 192.168.165.244
2019-11-20T16:47:57.872Z [DEBUG]  Reconciling ENI/IP pool info...
2019-11-20T16:47:57.873Z [DEBUG]  Total number of interfaces found: 3
2019-11-20T16:47:57.874Z [DEBUG]  Found ENI: eni-02797faf7a67b1590, MAC 06:48:a2:a8:91:f0, device 3
2019-11-20T16:47:57.877Z [DEBUG]  Found ENI: eni-0ea92c15e054d06d3, MAC 06:91:7d:d1:10:9c, device 2
2019-11-20T16:47:57.880Z [DEBUG]  Found ENI: eni-0578f3df9b86e409c, MAC 06:f1:59:05:99:ac, device 0
2019-11-20T16:47:57.882Z [DEBUG]  Adding ENI(eni-02797faf7a67b1590)'s IPv4 address 192.168.165.244 to datastore
2019-11-20T16:47:57.882Z [DEBUG]  Reconciled IP 192.168.165.244 on ENI eni-02797faf7a67b1590

***** 3) k8s asks ipamd to add a new network for pod fluentd-cloudwatch-chh96
***** ipamd assigns IP 192.168.165.244 from eni-02797faf7a67b1590
2019-11-20T16:48:00.711Z [INFO] Received AddNetwork for NS /proc/11349/ns/net, Pod fluentd-cloudwatch-chh96, NameSpace amazon-cloudwatch, Container e415008a45cc9f34eed550c837c69229df80dc3c2a6ad45dab0c9309d1d3f960, ifname eth0
2019-11-20T16:48:00.711Z [DEBUG]  AssignIPv4Address: IP address pool stats: total: 42, assigned 7
2019-11-20T16:48:00.711Z [INFO] AssignPodIPv4Address: Assign IP 192.168.165.244 to pod (name fluentd-cloudwatch-chh96, namespace amazon-cloudwatch container e415008a45cc9f34eed550c837c69229df80dc3c2a6ad45dab0c9309d1d3f960)
2019-11-20T16:48:00.711Z [INFO] Send AddNetworkReply: IPv4Addr 192.168.165.244, DeviceNumber: 3, err: <nil>

***** 4) ipamd updates pod fluentd-cloudwatch-chh96 from k8s, with the IP assigned above in (3)
2019-11-20T16:48:27.106Z [INFO] Add/Update for Pod fluentd-cloudwatch-chh96 on my node, namespace = amazon-cloudwatch, IP = 192.168.165.244

***** 5) ipamd reconciles 2 ENIs from AWS local metadata, not including eni-02797faf7a67b1590
2019-11-20T16:48:57.946Z [DEBUG]  Reconciling ENI/IP pool info...
2019-11-20T16:48:57.958Z [DEBUG]  Total number of interfaces found: 2
2019-11-20T16:48:57.965Z [DEBUG]  Found ENI: eni-0ea92c15e054d06d3, MAC 06:91:7d:d1:10:9c, device 2
2019-11-20T16:48:57.977Z [DEBUG]  Found ENI: eni-0578f3df9b86e409c, MAC 06:f1:59:05:99:ac, device 0

***** THINGS GO BAD FROM HERE

***** 6) hereafter ipamd won't delete detached eni-02797faf7a67b1590 because it has pods assigned
2019-11-20T16:48:57.981Z [INFO] Reconcile and delete detached ENI eni-02797faf7a67b1590
2019-11-20T16:48:57.981Z [ERROR]  IP pool reconcile: Failed to delete ENI during reconcile: datastore: ENI is used and can not be deleted
2019-11-20T16:49:58.004Z [ERROR]  IP pool reconcile: Failed to delete ENI during reconcile: datastore: ENI is used and can not be deleted
2019-11-20T16:50:58.017Z [ERROR]  IP pool reconcile: Failed to delete ENI during reconcile: datastore: ENI is used and can not be deleted
2019-11-20T16:51:58.032Z [ERROR]  IP pool reconcile: Failed to delete ENI during reconcile: datastore: ENI is used and can not be deleted

***** 7) in addition ipamd continues to assign IPs from eni-02797faf7a67b1590 (device 3)
2019-11-20T19:01:37.043Z [INFO] Send AddNetworkReply: IPv4Addr 192.168.167.47, DeviceNumber: 3, err: <nil>
```

--- a/ipamd/datastore/data_store.go
+++ b/ipamd/datastore/data_store.go
@@ -78,6 +78,18 @@ var (
 			Help: "The number of IP addresses assigned to pods",
 		},
 	)
+	forceRemovedENIs = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "awscni_force_removed_enis",
+			Help: "The number of ENIs force removed while they had assigned pods",
+		},
+	)
+	forceRemovedIPs = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "awscni_force_removed_ips",
+			Help: "The number of IPs force removed while they had assigned pods",
+		},
+	)
 	prometheusRegistered = false
 )
 
@@ -146,6 +158,8 @@ func prometheusRegister() {
 		prometheus.MustRegister(enis)
 		prometheus.MustRegister(totalIPs)
 		prometheus.MustRegister(assignedIPs)
+		prometheus.MustRegister(forceRemovedENIs)
+		prometheus.MustRegister(forceRemovedIPs)
 		prometheusRegistered = true
 	}
 }
@@ -209,7 +223,7 @@ func (ds *DataStore) AddIPv4AddressToStore(eniID string, ipv4 string) error {
 }
 
 // DelIPv4AddressFromStore delete an IP of ENI from datastore
-func (ds *DataStore) DelIPv4AddressFromStore(eniID string, ipv4 string) error {
+func (ds *DataStore) DelIPv4AddressFromStore(eniID string, ipv4 string, force bool) error {
 	ds.lock.Lock()
 	defer ds.lock.Unlock()
 	log.Debugf("Deleting ENI(%s)'s IPv4 address %s from datastore", eniID, ipv4)
@@ -226,7 +240,18 @@ func (ds *DataStore) DelIPv4AddressFromStore(eniID string, ipv4 string) error {
 	}
 
 	if ipAddr.Assigned {
-		return errors.New(IPInUseError)
+		if !force {
+			return errors.New(IPInUseError)
+		}
+		log.Warnf("Force deleting assigned ip %s on eni %s", ipv4, eniID)
+		forceRemovedIPs.Inc()
+		decrementAssignedCount(ds, curENI, ipAddr)
+		for key, info := range ds.podsIP {
+			if info.IP == ipv4 {
+				delete(ds.podsIP, key)
+				break
+			}
+		}
 	}
 
 	ds.total--
@@ -303,6 +328,17 @@ func incrementAssignedCount(ds *DataStore, eni *ENIIPPool, addr *AddressInfo) {
 	ds.assigned++
 	eni.AssignedIPv4Addresses++
 	addr.Assigned = true
+	// Prometheus gauge
+	assignedIPs.Set(float64(ds.assigned))
+}
+
+func decrementAssignedCount(ds *DataStore, eni *ENIIPPool, addr *AddressInfo) {
+	ds.assigned--
+	eni.AssignedIPv4Addresses--
+	addr.Assigned = false
+	curTime := time.Now()
+	eni.lastUnassignedTime = curTime
+	addr.UnassignedTime = curTime
 	// Prometheus gauge
 	assignedIPs.Set(float64(ds.assigned))
 }
@@ -431,7 +467,7 @@ func (ds *DataStore) RemoveUnusedENIFromStore(warmIPTarget int, minimumIPTarget 
 }
 
 // RemoveENIFromDataStore removes an ENI from the datastore.  It return nil on success or an error.
-func (ds *DataStore) RemoveENIFromDataStore(eni string) error {
+func (ds *DataStore) RemoveENIFromDataStore(eni string, force bool) error {
 	ds.lock.Lock()
 	defer ds.lock.Unlock()
 
@@ -440,9 +476,26 @@ func (ds *DataStore) RemoveENIFromDataStore(eni string) error {
 		return errors.New(UnknownENIError)
 	}
 
-	// Only unused ENIs can be deleted
-	if eniIPPool.AssignedIPv4Addresses != 0 {
-		return errors.New(ENIInUseError)
+	if eniIPPool.hasPods() {
+		if !force {
+			return errors.New(ENIInUseError)
+		}
+		// This scenario can occur if the reconciliation process discovered this eni was detached
+		// from the EC2 instance outside of the control of ipamd.  If this happens, there's nothing
+		// we can do other than force all pods to be unassigned from the IPs on this eni.
+		log.Warnf("Force removing eni %s with %d assigned pods", eni, eniIPPool.AssignedIPv4Addresses)
+		forceRemovedENIs.Inc()
+		forceRemovedIPs.Add(float64(eniIPPool.AssignedIPv4Addresses))
+		for _, addr := range eniIPPool.IPv4Addresses {
+			if addr.Assigned {
+				decrementAssignedCount(ds, eniIPPool, addr)
+			}
+		}
+		for key, info := range ds.podsIP {
+			if info.DeviceNumber == eniIPPool.DeviceNumber {
+				delete(ds.podsIP, key)
+			}
+		}
 	}
 
 	ds.total -= len(eniIPPool.IPv4Addresses)
@@ -478,13 +531,7 @@ func (ds *DataStore) UnassignPodIPv4Address(k8sPod *k8sapi.K8SPodInfo) (ip strin
 	for _, eni := range ds.eniIPPools {
 		ip, ok := eni.IPv4Addresses[ipAddr.IP]
 		if ok && ip.Assigned {
-			ip.Assigned = false
-			ds.assigned--
-			assignedIPs.Set(float64(ds.assigned))
-			eni.AssignedIPv4Addresses--
-			curTime := time.Now()
-			ip.UnassignedTime = curTime
-			eni.lastUnassignedTime = curTime
+			decrementAssignedCount(ds, eni, ip)
 			log.Infof("UnassignPodIPv4Address: pod (Name: %s, NameSpace %s Sandbox %s)'s ipAddr %s, DeviceNumber%d",
 				k8sPod.Name, k8sPod.Namespace, k8sPod.Sandbox, ip.Address, eni.DeviceNumber)
 			delete(ds.podsIP, podKey)

--- a/ipamd/datastore/data_store_test.go
+++ b/ipamd/datastore/data_store_test.go
@@ -67,7 +67,7 @@ func TestDeleteENI(t *testing.T) {
 	assert.Equal(t, len(eniInfos.ENIIPPools), 2)
 
 	// Add an IP and assign a pod.
-	err = ds.AddIPv4AddressFromStore("eni-1", "1.1.1.1")
+	err = ds.AddIPv4AddressToStore("eni-1", "1.1.1.1")
 	assert.NoError(t, err)
 	podInfo := &k8sapi.K8SPodInfo{
 		Name:      "pod-1",

--- a/ipamd/datastore/data_store_test.go
+++ b/ipamd/datastore/data_store_test.go
@@ -54,17 +54,37 @@ func TestDeleteENI(t *testing.T) {
 	eniInfos := ds.GetENIInfos()
 	assert.Equal(t, len(eniInfos.ENIIPPools), 3)
 
-	err = ds.RemoveENIFromDataStore("eni-2")
+	err = ds.RemoveENIFromDataStore("eni-2", false)
 	assert.NoError(t, err)
 
 	eniInfos = ds.GetENIInfos()
 	assert.Equal(t, len(eniInfos.ENIIPPools), 2)
 
-	err = ds.RemoveENIFromDataStore("unknown-eni")
+	err = ds.RemoveENIFromDataStore("unknown-eni", false)
 	assert.Error(t, err)
 
 	eniInfos = ds.GetENIInfos()
 	assert.Equal(t, len(eniInfos.ENIIPPools), 2)
+
+	// Add an IP and assign a pod.
+	err = ds.AddIPv4AddressFromStore("eni-1", "1.1.1.1")
+	assert.NoError(t, err)
+	podInfo := &k8sapi.K8SPodInfo{
+		Name:      "pod-1",
+		Namespace: "ns-1",
+		IP:        "1.1.1.1",
+	}
+	ip, device, err := ds.AssignPodIPv4Address(podInfo)
+	assert.NoError(t, err)
+	assert.Equal(t, "1.1.1.1", ip)
+	assert.Equal(t, 1, device)
+
+	// Test force removal.  The first call fails because eni-1 has an IP with a pod assigned to it,
+	// but the second call force-removes it and succeeds.
+	err = ds.RemoveENIFromDataStore("eni-1", false)
+	assert.Error(t, err)
+	err = ds.RemoveENIFromDataStore("eni-1", true)
+	assert.NoError(t, err)
 }
 
 func TestAddENIIPv4Address(t *testing.T) {
@@ -158,16 +178,39 @@ func TestDelENIIPv4Address(t *testing.T) {
 	assert.Equal(t, ds.total, 3)
 	assert.Equal(t, len(ds.eniIPPools["eni-1"].IPv4Addresses), 3)
 
-	err = ds.DelIPv4AddressFromStore("eni-1", "1.1.1.2")
+	err = ds.DelIPv4AddressFromStore("eni-1", "1.1.1.2", false)
 	assert.NoError(t, err)
 	assert.Equal(t, ds.total, 2)
 	assert.Equal(t, len(ds.eniIPPools["eni-1"].IPv4Addresses), 2)
 
 	// delete a unknown IP
-	err = ds.DelIPv4AddressFromStore("eni-1", "10.10.10.10")
+	err = ds.DelIPv4AddressFromStore("eni-1", "10.10.10.10", false)
 	assert.Error(t, err)
 	assert.Equal(t, ds.total, 2)
 	assert.Equal(t, len(ds.eniIPPools["eni-1"].IPv4Addresses), 2)
+
+	// Assign a pod.
+	podInfo := &k8sapi.K8SPodInfo{
+		Name:      "pod-1",
+		Namespace: "ns-1",
+		IP:        "1.1.1.1",
+	}
+	ip, device, err := ds.AssignPodIPv4Address(podInfo)
+	assert.NoError(t, err)
+	assert.Equal(t, "1.1.1.1", ip)
+	assert.Equal(t, 1, device)
+
+	// Test force removal.  The first call fails because the IP has a pod assigned to it, but the
+	// second call force-removes it and succeeds.
+	err = ds.DelIPv4AddressFromStore("eni-1", "1.1.1.1", false)
+	assert.Error(t, err)
+	assert.Equal(t, ds.total, 2)
+	assert.Equal(t, len(ds.eniIPPools["eni-1"].IPv4Addresses), 2)
+
+	err = ds.DelIPv4AddressFromStore("eni-1", "1.1.1.1", true)
+	assert.NoError(t, err)
+	assert.Equal(t, ds.total, 1)
+	assert.Equal(t, len(ds.eniIPPools["eni-1"].IPv4Addresses), 1)
 }
 
 func TestPodIPv4Address(t *testing.T) {

--- a/ipamd/ipamd.go
+++ b/ipamd/ipamd.go
@@ -552,8 +552,7 @@ func (c *IPAMContext) tryUnassignIPsFromAll() {
 			for _, toDelete := range ips {
 				// Don't force the delete, since a freeable IP might have been assigned to a pod
 				// before we get around to deleting it.
-				const force = false
-				err := c.dataStore.DelIPv4AddressFromStore(eniID, toDelete, force)
+				err := c.dataStore.DelIPv4AddressFromStore(eniID, toDelete, false /* force */)
 				if err != nil {
 					log.Warnf("Failed to delete IP %s on ENI %s from datastore: %s", toDelete, eniID, err)
 					ipamdErrInc("decreaseIPPool")
@@ -996,8 +995,7 @@ func (c *IPAMContext) nodeIPPoolReconcile(interval time.Duration) {
 		log.Infof("Reconcile and delete detached ENI %s", eni)
 		// Force the delete, since aws local metadata has told us that this ENI is no longer
 		// attached, so any IPs assigned from this ENI will no longer work.
-		const force = true
-		err = c.dataStore.RemoveENIFromDataStore(eni, force)
+		err = c.dataStore.RemoveENIFromDataStore(eni, true /* force */)
 		if err != nil {
 			log.Errorf("IP pool reconcile: Failed to delete ENI during reconcile: %v", err)
 			ipamdErrInc("eniReconcileDel")
@@ -1071,8 +1069,7 @@ func (c *IPAMContext) eniIPPoolReconcile(ipPool map[string]*datastore.AddressInf
 		log.Debugf("Reconcile and delete IP %s on ENI %s", existingIP, eni)
 		// Force the delete, since aws local metadata has told us that this ENI is no longer
 		// attached, so any IPs assigned from this ENI will no longer work.
-		const force = true
-		err := c.dataStore.DelIPv4AddressFromStore(eni, existingIP, force)
+		err := c.dataStore.DelIPv4AddressFromStore(eni, existingIP, true /* force */)
 		if err != nil {
 			log.Errorf("Failed to reconcile and delete IP %s on ENI %s, %v", existingIP, eni, err)
 			ipamdErrInc("ipReconcileDel")

--- a/ipamd/ipamd.go
+++ b/ipamd/ipamd.go
@@ -550,7 +550,10 @@ func (c *IPAMContext) tryUnassignIPsFromAll() {
 			// Delete IPs from datastore
 			var deletedIPs []string
 			for _, toDelete := range ips {
-				err := c.dataStore.DelIPv4AddressFromStore(eniID, toDelete)
+				// Don't force the delete, since a freeable IP might have been assigned to a pod
+				// before we get around to deleting it.
+				const force = false
+				err := c.dataStore.DelIPv4AddressFromStore(eniID, toDelete, force)
 				if err != nil {
 					log.Warnf("Failed to delete IP %s on ENI %s from datastore: %s", toDelete, eniID, err)
 					ipamdErrInc("decreaseIPPool")
@@ -991,7 +994,10 @@ func (c *IPAMContext) nodeIPPoolReconcile(interval time.Duration) {
 	// Sweep phase: since the marked ENI have been removed, the remaining ones needs to be sweeped
 	for eni := range curENIs.ENIIPPools {
 		log.Infof("Reconcile and delete detached ENI %s", eni)
-		err = c.dataStore.RemoveENIFromDataStore(eni)
+		// Force the delete, since aws local metadata has told us that this ENI is no longer
+		// attached, so any IPs assigned from this ENI will no longer work.
+		const force = true
+		err = c.dataStore.RemoveENIFromDataStore(eni, force)
 		if err != nil {
 			log.Errorf("IP pool reconcile: Failed to delete ENI during reconcile: %v", err)
 			ipamdErrInc("eniReconcileDel")
@@ -1063,7 +1069,10 @@ func (c *IPAMContext) eniIPPoolReconcile(ipPool map[string]*datastore.AddressInf
 	// Sweep phase, delete remaining IPs
 	for existingIP := range ipPool {
 		log.Debugf("Reconcile and delete IP %s on ENI %s", existingIP, eni)
-		err := c.dataStore.DelIPv4AddressFromStore(eni, existingIP)
+		// Force the delete, since aws local metadata has told us that this ENI is no longer
+		// attached, so any IPs assigned from this ENI will no longer work.
+		const force = true
+		err := c.dataStore.DelIPv4AddressFromStore(eni, existingIP, force)
 		if err != nil {
 			log.Errorf("Failed to reconcile and delete IP %s on ENI %s, %v", existingIP, eni, err)
 			ipamdErrInc("ipReconcileDel")

